app/client/websocket/index.ts

```diff
@@ -1,220 +1,319 @@
-    token: string|null;
-    sequence: number;
-    connectFailCount: number;
-    eventCallback?: Function;
-    firstConnectCallback?: Function;
-    reconnectCallback?: Function;
-    errorCallback?: Function;
-    closeCallback?: Function;
-    connectingCallback?: Function;
-    stop: boolean;
-    connectionTimeout: any;
-
-    constructor() {
-        this.connectionUrl = '';
-        this.token = null;
-        this.sequence = 1;
+export default class WebSocketClient {
+    private conn?: WebSocketClientInterface;
+    private connectionTimeout: any;
+    private connectionId: string;
+    private token: string;
+
+    // responseSequence is the number to track a response sent
+    // via the websocket. A response will always have the same sequence number
+    // as the request.
+    private responseSequence: number;
+
+    // serverSequence is the incrementing sequence number from the
+    // server-sent event stream.
+    private serverSequence: number;
+    private connectFailCount: number;
+    private eventCallback?: Function;
+    private firstConnectCallback?: () => void;
+    private missedEventsCallback?: () => void;
+    private reconnectCallback?: () => void;
+    private errorCallback?: Function;
+    private closeCallback?: (connectFailCount: number, lastDisconnect: number) => void;
+    private connectingCallback?: () => void;
+    private stop: boolean;
+    private lastConnect: number;
+    private lastDisconnect: number;
+
+    private serverUrl: string;
+
+    constructor(serverUrl: string, token: string, lastDisconnect = 0) {
+        this.connectionId = '';
+        this.token = token;
+        this.responseSequence = 1;
+        this.serverSequence = 0;
         this.connectFailCount = 0;
         this.stop = false;
+        this.serverUrl = serverUrl;
+        this.lastConnect = 0;
+        this.lastDisconnect = lastDisconnect;
     }
 
-    initialize(token: string|null, opts = {}) {
+    public async initialize(opts = {}) {
         const defaults = {
             forceConnection: true,
-            connectionUrl: this.connectionUrl,
         };
 
-        const {connectionUrl, forceConnection, ...additionalOptions} = Object.assign({}, defaults, opts);
+        const {forceConnection} = Object.assign({}, defaults, opts);
 
         if (forceConnection) {
             this.stop = false;
         }
 
-        return new Promise((resolve, reject) => {
-            if (this.conn) {
-                resolve(null);
-                return;
-            }
+        if (this.conn) {
+            return;
+        }
 
-            if (connectionUrl == null) {
-                console.log('websocket must have connection url'); //eslint-disable-line no-console
-                reject(new Error('websocket must have connection url'));
-                return;
-            }
+        const database = DatabaseManager.serverDatabases[this.serverUrl]?.database;
+        if (!database) {
+            return;
```

Comment:
should this throw an error?

Reply:
Similar concern with the other websocket action. Not sure if desirable to stop the websocket client execution throwing an error. Let's talk to discuss this further.